[Docs] Conditional Notifications (#1696)
mckinsey/vizro: vizro-core/docs/pages/user-guides/conditional-notifications-actions.md

# How to use conditional notifications

> This guide shows you how to use **conditional notifications** that show automatically based on a [custom action's](custom-actions.md) outcome. Instead of manually calling [`show_notification`][vizro.actions.show_notification] or [`update_notification`][vizro.actions.update_notification], you define a `notifications` argument as a dictionary in the [`vm.Action`][vizro.models.Action] upfront, and Vizro dispatches the right notification for you.
> 
> This section explains how to handle **success** and **error** outcomes of custom actions, show **progress** indicators while actions run, define custom notification keys for more specific scenarios, and template notification text with runtime values.

Reviewer: I find this a bit hard to digest because at this point I don't fully understand what conditional notifications are 🤔 and the details given here jump into technical implementation immediately. Curious to hear @stichbury thoughts. But I would probably explain first what we understand under "conditional notifications" and then high-level explain what this user guide shows.

```suggestion
Conditional notifications are messages that are displayed automatically based on the outcome of an action. For example, a notification can indicate whether an action was completed successfully or failed.

This guide describes how conditional notifications work with [custom actions](https://github.com/mckinsey/vizro/pull/1696/custom-actions.md)
 and how they can be used to provide users with feedback at the right moment.

The following sections cover how conditional notifications can represent different outcomes, such as success and error states, display progress indicators while actions are running, and support more specific scenarios and dynamic content.
```

Reviewer 2: I found the first paragraph very hard to get my head around as there's explanation there, plus "how to...". @huong-li-nguyen's edits definitely improve things, thank you 🙏 

@petar-qb I would add some more copy to the custom actions section just to explain that conditional notifications are a thing, and you can take your text from here, over into the appropriate custom actions section of docs:

> Instead of manually calling [`show_notification`][vizro.actions.show_notification] or [`update_notification`][vizro.actions.update_notification], you define a `notifications` argument as a dictionary in the [`vm.Action`][vizro.models.Action] upfront, and Vizro dispatches the right notification for you.

## Conditional notifications for custom actions

> This section covers actions defined with `@capture("action")`. If you are new to custom actions, see [How to create custom actions](custom-actions.md) first.
> 
> ### Success and error notifications
> 
> The two most common built-in keys are `"success"` and `"error"`. If a `"success"` notification is defined, it shows when the action returns normally (without raising an exception), and if an `"error"` notification is defined, it shows when the action raises an exception.

Reviewer: ```suggestion
This section focuses on conditional notifications used with actions defined via `@capture("action")`. If you're not familiar with custom actions, see [How to create custom actions](custom-actions.md) first.

### Success and error notifications

Two common types of conditional notifications are **"success"** and **"error"**. A **"success"** notification is shown when an action completes as expected. An **"error"** notification is shown automatically if the action raises an exception, indicating that something went wrong during execution.
```

!!! example "Success and error notifications"

    === "app.py"

        ```{.python hl_lines="8-12 24-27"}
        import random

        import vizro.models as vm
        from vizro import Vizro
        from vizro.models.types import capture

        @capture("action")
        def run_pipeline():
            if random.random() < 0.5:
                raise Exception("Random error.")
            return "Metrics updated"

        page = vm.Page(
            title="Success and error notifications",
            layout=vm.Flex(),
            components=[
                vm.Button(
                    text="Run pipeline",
                    actions=vm.Action(
                        function=run_pipeline(),
                        outputs="pipeline_output",
                        notifications={
                            "success": "Pipeline completed.",
                            "error": "Pipeline failed.",
                        },
                    ),
                ),
                vm.Text(id="pipeline_output", text="Click the button to run the pipeline."),
            ],
        )

> 

Reviewer: Could you add code annotations inside these examples? :) 

Like 
1: Define custom action that raises an exception
2: Define messages for keys success and error etc. 
3: XXXX

See our docs examples in other places

        dashboard = vm.Dashboard(pages=[page])
        Vizro().build(dashboard).run()
        ```

### Progress notifications

> Defined `"progress"` notification shows **before** the action starts and is automatically replaced by the success or error notification once the action completes. This is useful for long-running operations where you want to show a loading indicator.

Reviewer: ```suggestion
A **"progress"** notification is shown while an action is running. It appears before the action starts and is automatically replaced by a success or error notification once the action completes.

This is useful for long-running actions where you want to indicate that work is in progress and provide feedback to the user.
```

!!! example "Progress notification"

    === "app.py"

        ```{.python hl_lines="11 27"}
        import random
        from time import sleep

        import vizro.models as vm
        from vizro import Vizro
        from vizro.models.types import capture

>         @capture("action")

Reviewer: Same comment here with code annotations

        def run_pipeline():
            sleep(random.uniform(1, 2))
            if random.random() < 0.5:
                raise Exception("Random error.")
            return "Metrics updated"

        page = vm.Page(
            title="Progress notification",
            layout=vm.Flex(),
            components=[
                vm.Button(
                    text="Run pipeline",
                    actions=vm.Action(
                        function=run_pipeline(),
                        outputs="pipeline_output",
                        notifications={
                            "progress": "Running pipeline...",
                            "success": "Pipeline completed.",
                            "error": "Pipeline failed.",
                        },
                    ),
                ),
                vm.Text(id="pipeline_output", text="Click the button to run the pipeline."),
            ],
        )

        dashboard = vm.Dashboard(pages=[page])
        Vizro().build(dashboard).run()
        ```

### Custom notification keys

> Beyond the three built-in keys (`progress`, `success`, `error`), you can add **custom keys** to `notifications`. These custom notifications are not shown automatically though, but they show only when the action function explicitly **returns** or **raises** with a matching key.

Reviewer: ```suggestion
In addition to the built-in keys (`progress`, `success`, `error`), you can define your own custom notification keys. These allow you to handle more specific outcomes beyond just `success` or `failure`.

Custom notifications are only shown when the action explicitly selects them. This happens when the action function either returns a key or raises an exception with a key that matches one defined in notifications.
```

> To select a custom notification on success, append the key after the return value:

Reviewer: ```suggestion
To show a custom notification on success, return the key alongside the result:
```

```python
return result_text, "pipeline_partial_success"
```

> To select a custom notification on error, pass the key as a second argument to the exception:

Reviewer: ```suggestion
To show a custom notification on error, include the key when raising an exception:
```

```python
raise Exception("Pipeline failed.", "pipeline_partial_error")
```

> The key must match an entry in the defined `notifications` dictionary. If no key is returned from the action, the `"success"` notification shows on return (if defined), and the `"error"` notification shows on exception (if defined).
> 
> Custom keys default to `variant="info"` unless overridden with a `show_notification` object (see [Full notification customization](#full-notification-customization)).

Reviewer: ```suggestion
If no custom key is provided, the default behavior applies: a "success" notification is shown when the action completes, and an "error" notification is shown if an exception is raised (if these are defined).

Custom notifications use the "info" style by default, unless configured otherwise (see our user guide on [customizing notifications](#full-notification-customization)).
```

> !!! note
> 
>     You can explicitly return built-in keys like `"success"` or `"error"` from your action function, but this is rarely needed since they are handled by default.
> 

Reviewer: ```suggestion
```
I would just remove that part :)

!!! example "Custom notification keys"

    === "app.py"

        ```{.python hl_lines="17 21 37-38"}
        import random
        from time import sleep

        import vizro.models as vm
        from vizro import Vizro
        from vizro.models.types import capture

        @capture("action")

>         def run_pipeline():

Reviewer: Same thing here on adding code annotations

            if roll < 0.75:
                return "Metrics updated"
            if roll <= 1:
                return "Metrics updated", "pipeline_partial_success"

        page = vm.Page(
            title="Custom notification keys",
            layout=vm.Flex(),
            components=[
                vm.Button(
                    text="Run pipeline",
                    actions=vm.Action(
                        function=run_pipeline(),
                        outputs="pipeline_output",
                        notifications={
                            "progress": "Running pipeline...",
                            "success": "Pipeline completed.",
                            "error": "Pipeline failed.",
                            "pipeline_partial_success": "Pipeline partially completed.",
                            "pipeline_partial_error": "Pipeline partially failed.",
                        },
                    ),
                ),
                vm.Text(id="pipeline_output", text="Click the button to run the pipeline."),
            ],
        )

        dashboard = vm.Dashboard(pages=[page])
        Vizro().build(dashboard).run()
        ```

> ### Templating notification text

Reviewer: ```suggestion
### Dynamic notification content
```

> Notification messages support **template variables** that are replaced with action's runtime values:
> 
> - **`{{result}}`**: template within the notification's text that's replaced with an extra value that's returned from the action's function as the second element of a tuple.
> 
>     - Example for return values: `return output, ("key", "detail text")`.
>     - Example for exceptions: `raise Exception("error message", ("key", "detail text"))`.
> 
>     The second element of the tuples replace the `{{result}}` textual template. Resolves to an empty string if not provided.
> 
> - **`{{error_msg}}`**: template within the notification's text that's replaced with the first argument of the exception. Resolves to an empty string if not provided.

Reviewer: I thought there were too many technical details again, which I would understand as a user. So in simple terms this is what I would understand. Feel free to correct:

```suggestion
Notification messages can include **template variables** that are filled in with values from the action at runtime. This allows you to display dynamic information, such as results or error details, directly in the notification.

Two templates are supported:

- `**{{result}}**`: Replaced with additional information returned by the action. This value is optional and can be included alongside the main result when the action completes or raises an exception. If no value is provided, it resolves to an empty string.
- `**{{error_msg}}:**` Replaced with the error message from an exception. If no error message is available, it resolves to an empty string.

These templates make it possible to provide more informative and contextual feedback to users without hardcoding the message content.
```

!!! example "Templating with `{{result}}` and `{{error_msg}}`"

    === "app.py"

        ```{.python hl_lines="17 19 21 23 37-40"}
        import random
        from time import sleep

        import vizro.models as vm
        from vizro import Vizro
        from vizro.models.types import capture

        @capture("action")
        def run_pipeline():
            duration = random.uniform(1, 2)
            duration_result = f"Duration: {duration:.1f}s"
            sleep(duration)

            roll = random.random()
            if roll < 0.25:
                raise Exception("Random error.", ("error", duration_result))
            if roll < 0.5:
                raise Exception("Random error.", ("pipeline_partial_error", duration_result))
            if roll < 0.75:
                return "Metrics updated", ("success", duration_result)

>             if roll <= 1:

Reviewer: You can then add more technical details in the code annotations here. Maybe do instructions again:

1. XX
2. XX
3. XX

                return "Metrics updated", ("pipeline_partial_success", duration_result)

        page = vm.Page(
            title="Templated notifications",
            layout=vm.Flex(),
            components=[
                vm.Button(
                    text="Run pipeline",
                    actions=vm.Action(
                        function=run_pipeline(),
                        outputs="pipeline_output",
                        notifications={
                            "progress": "Running pipeline...",
                            "success": "Pipeline done. {{result}}",
                            "error": "Pipeline failed: {{error_msg}} \n {{result}}",
                            "pipeline_partial_success": "Pipeline partially completed. {{result}}",
                            "pipeline_partial_error": "Pipeline partially failed: {{error_msg}} \n {{result}}",
                        },
                    ),
                ),
                vm.Text(id="pipeline_output", text="Click the button to run the pipeline."),
            ],
        )

        dashboard = vm.Dashboard(pages=[page])
        Vizro().build(dashboard).run()
        ```

> The `{{result}}` template is populated from the **second** element of the tuple passed alongside a return value or exception. It's `duration_result"` in the example above.
> 
> The `{{error_msg}}` template is populated from the **first** argument of the exception. It's `"Random error."` in the example above.

Reviewer: ```suggestion
In this example, `**{{result}}**` is filled with the additional detail returned by the action (for example, the duration of the pipeline run).

The `**{{error_msg}}**` template is filled with the error message from the exception (for example, "Random error.").
```

> ### Templating progress text with action inputs

Reviewer: ```suggestion
### Using action inputs in progress notifications
```

> Progress notification text can also be templated with **action runtime input arguments** referenced by their parameter names `{{param_name}}`.

Reviewer: ```suggestion
Progress notifications can also include template variables based on the action's input values. These variables are referenced using the parameter names, for example `{{param_name}}`.

This allows the progress message to reflect the current inputs of the action, making it more informative. For example, you can display the number of retries configured while the action is running.
```

!!! example "Dynamic progress text"

    === "app.py"

>         ```{.python hl_lines="12 35 39 42"}

Reviewer: Code annotations :)

                vm.Button(
                    text="Run pipeline",
                    actions=vm.Action(
                        function=run_pipeline(max_retries="slider_id"),
                        outputs="pipeline_output",
                        notifications={
                            "progress": "Running pipeline... (max retries: {{max_retries}})...",
                            "success": "Pipeline done. {{result}}",
                            "error": "Pipeline failed: {{error_msg}} - {{result}}",
                            "pipeline_partial_success": "Pipeline partially completed. {{result}}",
                            "pipeline_partial_error": "Pipeline partially failed: {{error_msg}} - {{result}}",
                            "pipeline_max_retries_error": "Pipeline failed after {{result}} retries.",
                        },
                    ),
                ),
                vm.Text(id="pipeline_output", text="Click the button to run the pipeline."),
            ],
        )

        dashboard = vm.Dashboard(pages=[page])
        Vizro().build(dashboard).run()
        ```

> ### Full notification customization

Reviewer: ```suggestion
### Customizing notification appearance and behavior
```

> Instead of plain text, a notification dictionary value can be a [`show_notification`][vizro.actions.show_notification] or [`update_notification`][vizro.actions.update_notification] model, giving full control over `variant`, `title`, `icon`, `auto_close` and more. See the [notifications guide](notification-actions.md) for all available parameters. Template variables like `{{result}}` still work inside the `text` argument. This might be useful when you want to change the default `variant="info"` for custom keys.

Reviewer: ```suggestion
Notifications don't have to be limited to plain text. You can customize how they look and behave, for example, by adjusting the variant, adding a title or icon, or controlling whether they close automatically. See the [notifications guide](notification-actions.md) for a full list of available options.

To do this, use the [show_notification][vizro.actions.show_notification] or [update_notification][vizro.actions.update_notification] models instead of plain text when defining notifications. This allows notifications to be tailored to different scenarios, such as highlighting important errors or distinguishing partial successes from full successes. Template variables like {{result}} can still be used within customized notification text.
```

                duration_result = f"Duration: {duration:.1f}s"
                sleep(duration)

                roll = random.uniform(0, 5)
                if roll < 0.25:
                    raise Exception("Random error.", ("error", duration_result))
                if roll < 0.5:
                    raise Exception("Random error.", ("pipeline_partial_error", duration_result))
                if roll < 0.75:
                    return "Metrics updated", ("success", duration_result)
                if roll <= 1:
                    return "Metrics updated", ("pipeline_partial_success", duration_result)

            raise Exception("Random error.", ("pipeline_max_retries_error", max_retries))

        page = vm.Page(
            title="Templated notifications",
            layout=vm.Flex(),
            components=[
                vm.Slider(id="slider_id", min=1, max=3, step=1, value=1, title="Max retries"),
                vm.Button(
                    text="Run pipeline",
                    actions=vm.Action(
                        function=run_pipeline(max_retries="slider_id"),
                        outputs="pipeline_output",
                        notifications={

>                             "progress": va.show_notification(id="progress_id", variant="progress", text="Running pipeline... (max retries: {{max_retries}})..."),

Reviewer: Add code annotations here too

                            "success": "Pipeline done. {{result}}",
                            "error": "Pipeline failed: {{error_msg}} - {{result}}",
                            "pipeline_partial_success": va.update_notification(variant="success", notification="progress_id", text="Pipeline partially completed. {{result}}"),
                            "pipeline_partial_error": va.update_notification(variant="error", notification="progress_id", text="Pipeline partially failed: {{error_msg}} - {{result}}"),
                            "pipeline_max_retries_error": va.update_notification(variant="error", notification="progress_id", text="Pipeline failed after {{result}} retries."),
                        },
                    ),
                ),
                vm.Text(id="pipeline_output", text="Click the button to run the pipeline."),
            ],
        )

        dashboard = vm.Dashboard(pages=[page])
        Vizro().build(dashboard).run()
        ```

> ### Debugging and error handling

Reviewer: I don't know if this should be an extra section or just an annotation?

Author: I addressed [your suggestion](https://github.com/mckinsey/vizro/pull/1696#discussion_r3092270328) and left the section. 

Interested what @antonymilne thinks, whether should we replace an entire section with a few annotations and/or notes.

> Every custom action has a default `notifications={"error": "Action failed."}`. You can override this with a more specific message or set it to `None` to disable it entirely.
> 
> When the error notification is disabled (`notifications={"error": None}`) and the app runs in debug mode (`Vizro().build(dashboard).run(debug=True)`), unhandled exceptions are forwarded to the [Dash Dev Tools](https://dash.plotly.com/devtools) debugger. When an `"error"` notification **is** defined, the exception is caught and shown as an in-app notification instead.
> 
> All exceptions (except `PreventUpdate`) are always logged to the server console, regardless of notification configuration.

Reviewer: ```suggestion
By default, every custom action shows a generic error notification ("Action failed.") if something goes wrong. You can override this with a more specific message, or disable error notifications entirely by setting `"error": None`.

If error notifications are disabled and the app is running in debug mode `(Vizro().build(dashboard).run(debug=True))`, any unhandled exceptions are shown in the [Dash Dev Tools](https://dash.plotly.com/devtools) debugger instead of in the app. If an "error" notification is defined, the exception is caught and displayed as a notification within the app.

All exceptions (except PreventUpdate) are always logged to the server console, regardless of how notifications are configured.
```

        page = vm.Page(
            title="Progress notification",
            layout=vm.Flex(),
            components=[
                vm.Button(
                    text="Run pipeline",
                    actions=vm.Action(
                        function=run_pipeline(),
                        outputs="pipeline_output",
                        notifications={
                            "progress": "Running pipeline...",  # (2)!
                            "success": "Pipeline completed.",
                            "error": "Pipeline failed.",
                        },
                    ),
                ),
                vm.Text(id="pipeline_output", text="Click the button to run the pipeline."),
            ],
        )

        dashboard = vm.Dashboard(pages=[page])
        Vizro().build(dashboard).run()
        ```

        1. Add a `sleep` call in the action function to simulate a long-running operation.
        1. A `"progress"` notification with the "Running pipeline..." text is shown immediately when the action starts, and then replaced by either the `"success"` or `"error"` notification once the action completes.

## Custom notification keys

> In addition to the built-in keys (`progress`, `success`, `error`), you can define your own custom notification keys to handle more specific outcomes beyond just `success` or `error`.

Reviewer 2: > ```In addition to the built-in keys (`progress`, `success`, `error`), you can define your own custom notification keys to handle more specific outcomes beyond just `success` or `error`. ```

In the section above, you use `"progress"` but here the double quotes are omitted. Is that OK?

        page = vm.Page(
            title="Templated notifications",
            layout=vm.Flex(),
            components=[
                vm.Button(
                    text="Run pipeline",
                    actions=vm.Action(
                        function=run_pipeline(),
                        outputs="pipeline_output",
                        notifications={
                            "progress": "Running pipeline...",
                            "success": "Pipeline done. {{result}}",  # (5)!
                            "error": "Pipeline failed: {{error_msg}} \n {{result}}",  # (6)!
                            "pipeline_partial_success": "Pipeline partially completed. {{result}}",  # (7)!
                            "pipeline_partial_error": "Pipeline partially failed: {{error_msg}} \n {{result}}",  # (8)!
                        },
                    ),
                ),
                vm.Text(id="pipeline_output", text="Click the button to run the pipeline."),
            ],
        )

        dashboard = vm.Dashboard(pages=[page])
        Vizro().build(dashboard).run()
        ```

>         1. Propagating dynamic value `duration_result` to the `"error"` notification text that replaces the `{{results}}` template.

Reviewer 2: Above, you use `{{result}}` rather than `{{results}}` so is this OK? You use the plural 4 times in this section.
